lib/models/StreamSplitter.js

```diff
@@ -0,0 +1,131 @@
+}
+
+class FileStreamSplitter extends stream.Writable {
+    constructor(options, { maxChunkSize, directory }) {
+        super(options);
+
+        this.maxChunkSize = maxChunkSize;
+
+        this.currentChunkPath = null;
+        this.currentChunkSize = null;
+        this.fileDescriptor = null;
+
+        this.directory = directory;
+        this.filenameTemplate = randomString(10);
+
+        this.part = 0;
+
+        this.on('error', this._finishChunk.bind(this));
+    }
+
+    _write(chunk, encoding, callback) {
+        Promise.resolve()
+            .then(() => {
+                // In order to start writing a chunk, we must first create
+                // a file system reference for it
+                if (this.fileDescriptor === null) {
+                    return this._newChunk();
+                }
+                return undefined;
+            })
+            .then(() => {
+                const overflow = this.currentChunkSize + chunk.length - this.maxChunkSize;
+
+                // If the chunk is bigger than the defined max chunk size,
+                // we need two passes to process the chunk
+                if (overflow > 0) {
+                    return this._writeChunk(chunk.slice(0, chunk.length - overflow))
+                        .then(this._finishChunk.bind(this))
+                        .then(this._newChunk.bind(this))
+                        .then(() => {
+                            return this._writeChunk(
+                                chunk.slice(chunk.length - overflow, chunk.length)
+                            );
+                        })
+                        .then(() => callback())
+                        .catch(callback);
+                }
+
+                // The chunk fits in the max chunk size
+                return this._writeChunk(chunk)
+                    .then(() => callback())
+                    .catch(callback);
+            })
+            .catch(callback);
+    }
+
+    _final(callback) {
+        if (this.fileDescriptor === null) {
+            callback();
+        }
+        else {
+            this._finishChunk()
+                .then(() => callback())
+                .catch(callback);
+        }
+    }
+
+    _writeChunk(chunk) {
+        return new Promise((resolve, reject) => {
+            fs.write(this.fileDescriptor, chunk, (err) => {
+                if (err) {
+                    return reject(err);
+                }
+
+                this.currentChunkSize += chunk.length;
+                return resolve();
+            });
+        });
+    }
+
+    _finishChunk() {
+        if (this.fileDescriptor === null) {
+            return Promise.resolve();
+        }
+
+        return new Promise((resolve, reject) => {
+            fs.close(this.fileDescriptor, (err) => {
+                if (err) {
+                    return reject(err);
+                }
+
+                this.emit('chunkFinished', this.currentChunkPath);
```

Comment:
Will `emit` wait until all event listeners have been invoked? I am wondering if it is possible that somehow `chunkStarted` from `_newChunk` is invoked before chunkFinished?

Reply:
`chunkStarted` is emitted and _afterwards_ resolved. We won't end up in the next `then` (where `chunkFinished` happens) until that promise is resolved